keps/sig-node/3619-supplemental-groups-policy/README.md

```diff
@@ -0,0 +1,857 @@
+  containers:
+  - image: some-image # uid=1000 belongs to gid=60000 in the image
+    sh: ["id"]
+```
+Then, groups of the first process of the container will be :
+- `1000`: specified by `runAsGroup`
+- `50000`: specified by `supplementalGroups`
+- `60000`: defined in the image (<strong><em>Note: Kubernetes keep this</em></strong>)
+
+Thus, the container logs will be:
+
+```console
+uid=1000(...) gid=1000(...) groups=1000(...),50000,60000(...)
+```
+
+This KEP proposes a complementary field named `SupplementalGroupsPolicy` that enables users to customize to respect/ignore the group membership of the primary UID of the container defined in the container image when applying `SupplementalGroups` to the container processes. 
+
+## Motivation
+
+As described above, `SupplementalGroups` will be just added to the groups defined in the container image for the container's primary UID. When a cluster enforces some security policy for pods that protects the value of `SupplementalGroups`, the effect of its enforcement is limited, i.e., cluster users can easily bypass the policy enforcement just by using a custom image. If such a bypass happened, it would be unexpected behavior for most cluster administrators because the enforcement is almost useless. Moreover, the bypass will cause unexpected file access permission. In some use cases, the unexpected file access permission will be a security concern. For example, using `hostPath` volumes could be a severe problem because UID/GIDs matter in accessing files/directories in the volumes.
+
+However, Kubernetes provides no API surface to prevent this bypass, although it could sometimes lead to a security concern. Because the behavior is implemented in CRI implementations actually, To mitigate this, the cluster administrators will need to deploy a custom low-level container runtime(e.g., [pfnet-research/strict-supplementalgroups-container-runtime](https://github.com/pfnet-research/strict-supplementalgroups-container-runtime)) that modifies OCI container runtime spec(`config.json`) produced by CRI implementations (e.g., containerd, cri-o). A custom `RuntimeClass` would be introduced for it. Nevertheless, It would be an extra operational burden for cluster administrators.
+
+Thus, this KEP proposes to offer a new API field named `SupplementalGroupsPolicy` that customize to respect/ignore the group membership of the primary UID of the container defined in the container image when applying `SupplementalGroups` to the container processes. The new API allows cluster administrators to deploy security policies that protect the `SupplementalGroupsPolicy` field in the cluster to avoid the unexpected bypass of `SupplementalGroups` described above.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+To Provide a new API field to control exactly which groups the container process belongs to.
```

Comment:
```suggestion
- Work with Container Runtimes to provide a configuration to apply the fix or keep the existing behavior
- Ensure there is a clear steps documented for end users to detect if their workload is affected 
- (Optional) provide helper APIs and/or tooling to simplify the detection

To Provide a new API field to control exactly which groups the container process belongs to.
```

Reply:
I put off committing the suggestion right now. see: t.b.d.

Reply:
definitely, thank you!

Reply:
added in 166d1998b2e0ab3a992f814e8f1020efa2697ace. PTAL.